pkg/controllers/updaterun/initialization_integration_test.go

```diff
@@ -0,0 +1,758 @@
+				ObservedGeneration: policySnapshot.Generation,
+				Reason:             "scheduled",
+			})
+			Expect(k8sClient.Status().Update(ctx, policySnapshot)).Should(Succeed(), "failed to update the policy snapshot condition")
+
+			By("Creating a new clusterStagedUpdateRun")
+			Expect(k8sClient.Create(ctx, updateRun)).To(Succeed())
+
+			By("Validating the initialization failed")
+			Eventually(func() error {
+				if err := k8sClient.Get(ctx, updateRunNamespacedName, updateRun); err != nil {
+					return err
+				}
+				if updateRun.Status.PolicySnapshotIndexUsed != policySnapshot.Name {
+					return fmt.Errorf("updateRun status `PolicySnapshotIndexUsed` mismatch: got %s, want %s", updateRun.Status.PolicySnapshotIndexUsed, policySnapshot.Name)
+				}
+				if updateRun.Status.PolicyObservedClusterCount != -1 {
+					return fmt.Errorf("updateRun status `PolicyObservedClusterCount` mismatch: got %d, want %d", updateRun.Status.PolicyObservedClusterCount, -1)
+				}
+				return nil
+			}, timeout, interval).Should(Succeed(), "failed to update the updateRun status with policy snapshot details")
+		})
+	})
+
+	Context("Test collectScheduledClusters", func() {
+		BeforeEach(func() {
+			By("Creating a new clusteResourcePlacement")
+			Expect(k8sClient.Create(ctx, crp)).To(Succeed())
+
+			By("Creating scheduling policy snapshot")
+			Expect(k8sClient.Create(ctx, policySnapshot)).To(Succeed())
+
+			By("Set the latest policy snapshot condition as fully scheduled")
+			meta.SetStatusCondition(&policySnapshot.Status.Conditions, metav1.Condition{
+				Type:               string(placementv1beta1.PolicySnapshotScheduled),
+				Status:             metav1.ConditionTrue,
+				ObservedGeneration: policySnapshot.Generation,
+				Reason:             "scheduled",
+			})
+			Expect(k8sClient.Status().Update(ctx, policySnapshot)).Should(Succeed(), "failed to update the policy snapshot condition")
+		})
+
+		It("Should fail to initialize if there is no selected or to-be-deleted cluster", func() {
+			By("Creating a new clusterStagedUpdateRun")
+			Expect(k8sClient.Create(ctx, updateRun)).To(Succeed())
+
+			By("Validating the initialization failed")
+			validateFailedInitCondition(ctx, updateRun, "no scheduled or to-be-deleted clusterResourceBindings found")
+		})
+
+		It("Should not report error if there are only to-be-deleted clusters", func() {
+			By("Creating a to-be-deleted clusterResourceBinding")
+			binding := getTestClusterResourceBinding(policySnapshot.Name+"a", "cluster-0")
+			Expect(k8sClient.Create(ctx, binding)).To(Succeed())
+
+			By("Creating a new clusterStagedUpdateRun")
+			Expect(k8sClient.Create(ctx, updateRun)).To(Succeed())
+
+			By("Validating the initialization not failed due to no selected cluster")
+			// it should fail due to strategy not found
+			validateFailedInitCondition(ctx, updateRun, "referenced clusterStagedUpdateStrategy not found")
+		})
+
+		It("Should retry if the bindings are not in Scheduled or Bound state", func() {
```

Comment:
we should not retry, the code is not right

Reply:
yea you are right. Forgot to change the error msg. Fixed.